zebra-network/src/address_book.rs

```diff
@@ -422,7 +442,24 @@ impl AddressBook {
     }
 
     /// Returns metrics for the addresses in this address book.
+    /// Only for use in tests.
+    ///
+    /// # Correctness
+    ///
+    /// Use [`AddressBook::address_metrics_watcher().borrow()`] in production code,
+    /// to avoid deadlocks.
+    #[cfg(test)]
     pub fn address_metrics(&self, now: chrono::DateTime<Utc>) -> AddressMetrics {
+        self.address_metrics_internal(now)
+    }
+
+    /// Returns metrics for the addresses in this address book.
+    ///
+    /// # Correctness
+    ///
+    /// Use [`AddressBook::address_metrics_watcher().borrow()`] in production code,
+    /// to avoid deadlocks.
+    fn address_metrics_internal(&self, now: chrono::DateTime<Utc>) -> AddressMetrics {
```

Comment:
So this is to use by tests only ? if so, we will want to add a `#[cfg(test)]` and `Only for use in tests.` to the description.

If it not a function for tests only then the Correctness text may be wrong.

Reply:
This method needs to be called by other address book methods, so we can update the sender end of the metrics channel, and the address book metrics.

See: https://github.com/ZcashFoundation/zebra/pull/3244/files#diff-c25b8ba12213af17e87b8fc2b897b38cca9b910e8a29ed25bbc3d918097a10aeR496

```rust
let m = self.address_metrics_internal(chrono_now);
let _ = self.address_metrics_tx.send(m);
```

Calling this method is ok inside the address book, because the thread already holds the address book mutex. But it's a bit risky outside the address book, which is why we have the metrics channel.

```suggestion
    /// Returns metrics for the addresses in this address book.
    ///
    /// # Correctness
    ///
    /// External callers should use [`AddressBook::address_metrics_watcher().borrow()`]
    /// in production code, to avoid deadlocks.
    /// (Using the watch channel receiver does not lock the address book mutex.)
    fn address_metrics_internal(&self, now: chrono::DateTime<Utc>) -> AddressMetrics {
```

Reply:
Thank you for the clarification.